File: Matrix.java
public class Matrix
{
    final String FLOAT_PRECISION = "%.2f";

    final String E_0ROWS = "E_0ROWS";    // matrix can't have 0 rows
    final String E_0COLS = "E_0COLS";    // matrix can't have 0 columns
    final String E_ROUTB = "E_ROUTB";    // row index out of bounds
    final String E_COUTB = "E_COUTB";    // column index out of bounds
    final String E_INCMP = "E_INCMP";    // incompatible matrices
    final String E_NOSQR = "E_NOSQR";    // not a square matrix
    final String E_DETR0 = "E_DETR0";    // during inversion, determinant is 0

    public boolean coderr;

    public int rows;
    public int cols;
    public double[][] matrix;

    /**
     * Default constructor not allowed
     */
    private Matrix() { }

    /**
     * Create a new Matrix object
     *
     * @param matrix If DDA is known, pass the DDA
     *
     * @throws Exception matrix can't have 0 rows - E_0ROWS
     * @throws Exception matrix can't have 0 columns - E_0COLS
     */
    public Matrix (double[][] matrix) throws Exception
    {
        this.rows = matrix.length;
        this.cols = matrix[0].length;
        if (this.rows < 1) {
            this.throwException (this.E_0ROWS, "matrix can't have 0 rows");
        }
        if (this.cols < 1) {
            this.throwException (this.E_0COLS, "matrix can't have 0 rows");
        }
        this.matrix = matrix;
        this.coderr = true;
    }

    /**
     * Create a new Matrix object
     *
     * @param rows If DDA is unknown, pass no. of rows
     * @param cols If DDA is unknown, pass no. of cols
     *
     * @throws Exception matrix can't have 0 rows - E_0ROWS
     * @throws Exception matrix can't have 0 columns - E_0COLS
     */
    public Matrix (int rows, int cols) throws Exception
    {
        if (rows < 1) {
            this.throwException (this.E_0ROWS, "matrix can't have 0 rows");
        }
        if (cols < 1) {
            this.throwException (this.E_0COLS, "matrix can't have 0 rows");
        }
        this.rows = rows;
        this.cols = cols;
        this.coderr = true;
        this.matrix = new double [rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                this.matrix [i][j] = 0;
            }
        }
    }

    private void throwException (String errCode, String errMsg) throws Exception
    {
        if (this.coderr) {
            throw new Exception (errCode);
        } else {
            throw new Exception ("Matrix: " + errMsg);
        }
    }

    /**
     * Create a null matrix of given size
     *
     * @param n size of matrix
     *
     * @return Matrix a null matrix
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    public static Matrix O (int n) throws Exception
    {
        Matrix om = new Matrix (n, n);
        return om;
    }

    /**
     * Create a null matrix of given size
     *
     * @param rows rows of matrix
     * @param cols cols of matrix
     *
     * @return Matrix a null matrix
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    public static Matrix O (int rows, int cols) throws Exception
    {
        Matrix om = new Matrix (rows, cols);
        return om;
    }

    /**
     * Create a unit matrix of given size
     *
     * @param n size of matrix
     *
     * @return Matrix a unit matrix
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    public static Matrix I (int n) throws Exception
    {
        Matrix im = new Matrix (n, n);
        for (int i = 0; i < n; i++) {
            im.set (i + 1, i + 1, 1);
        }
        return im;
    }

    /**
     * Get an element of the matrix from an index
     *
     * @param i row wise position of element (starts from 1)
     * @param j column wise position of element (starts from 1)
     *
     * @return double The value at index i, j
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    public double get (int i, int j) throws Exception
    {
        i--;
        j--;
        if (i < 0 || i > this.rows) {
            this.throwException (this.E_ROUTB, "row index out of bounds");
        }
        if (j < 0 || j > this.cols) {
            this.throwException (this.E_COUTB, "column index out of bounds");
        }
        return this.matrix [i][j];
    }

    /**
     * Set an element of the matrix to an index
     *
     * @param i row wise position of element (starts from 1)
     * @param j column wise position of element (starts from 1)
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    public void set (int i, int j,  double value) throws Exception
    {
        i--;
        j--;
        if (i < 0 || i > this.rows) {
            this.throwException (this.E_ROUTB, "row index out of bounds");
        }
        if (j < 0 || j > this.cols) {
            this.throwException (this.E_COUTB, "column index out of bounds");
        }
        this.matrix [i][j] = value;
    }

    /*
     * Compares two matrices for equality
     *
     * @param mtb The matrix to compare to
     *
     * @return boolean true if equal
     */
    public boolean equals (Matrix mtb)
    {
        if (this.rows != mtb.rows || this.cols != mtb.cols) {
             return false;
        }
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                if (this.matrix [i][j] != mtb.matrix [i][j]) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Add two compatible matrices
     *
     * @param mtb The matrix to add
     *
     * @return Matrix The matrix of sums
     *
     * @throws Exception If matrices aren't compatible - E_INCMP
     */
    public Matrix add (Matrix mtb) throws Exception
    {
        if (this.rows != mtb.rows || this.cols != mtb.cols) {
            this.throwException (this.E_INCMP, "incompatible matrices for addition");
        }
        Matrix rslt = new Matrix (this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                rslt.matrix [i][j] = this.matrix [i][j] + mtb.matrix [i][j];
            }
        }
        return rslt;
    }

    /**
     * Subtract two compatible matrices
     *
     * @param mtb The matrix to subtract
     *
     * @return Matrix The matrix of differences
     *
     * @throws Exception If matrices aren't compatible - E_INCMP
     */
    public Matrix subtract (Matrix mtb) throws Exception
    {
        if (this.rows != mtb.rows || this.cols != mtb.cols) {
            this.throwException (this.E_INCMP, "incompatible matrices for subtraction");
        }
        Matrix rslt = new Matrix (this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                rslt.matrix [i][j] = this.matrix [i][j] - mtb.matrix [i][j];
            }
        }
        return rslt;
    }

    /**
     * Multiplies a matrix by a scalar
     *
     * @param scalar Scalar to multiply by
     *
     * @return Matrix The matrix of products
     */
    public Matrix scale (double scalar) throws Exception
    {
        Matrix rslt = new Matrix (this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                rslt.matrix [i][j] = scalar * this.matrix [i][j];
            }
        }
        return rslt;
    }

    /**
     * Multiplies two compatible matrices
     *
     * @param mtb The matrix to multiply by
     *
     * @return Matrix The matrix after multiplication
     *
     * @throws Exception If matrices aren't compatible - E_INCMP
     */
    public Matrix multiply (Matrix mtb) throws Exception
    {
        if (this.cols != mtb.rows) {
            this.throwException (this.E_INCMP, "incompatible matrices for multiplication");
        }
        int m = this.rows;
        int n = this.cols; // same
        n = mtb.rows;  // same
        int o = mtb.cols;
        Matrix rslt = new Matrix (m, o);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < o; j++) {
                // zeroing the sum without changing datatypes
                double sum = this.matrix [0][0] - this.matrix [0][0];
                for (int k = 0; k < n; k++) {
                    sum += this.matrix [i][k] * mtb.matrix [k][j];
                }
                rslt.matrix [i][j] = sum;
            }
        }
        return rslt;
    }

    /**
     * Calculate matrix to the power of index
     *
     * @param index Power of matrix
     *
     * @return Matrix The resulting matrix
     *
     * @throws Exception same as errors of Matrix.multiply method
     */
    public Matrix power (int index) throws Exception
    {
        Matrix result = this;
        for (int i = 0; i < index - 1; i++) {
            result = result.multiply (this);
        }
        return result;
    }

    /**
     * Excludes a row and s column and generates a sub matrix. Useful for calculating determinants and cofactor matrices.
     *
     * @param row The row to exclude (starts from 1)
     * @param col The column to exclude (starts from 1)
     *
     * @return Matrix The sub matrix
     *
     * @throws Exception row index out of bounds - E_ROUTB
     * @throws Exception column index out of bounds - E_COUTB
     */
    private Matrix excludeRowCol (int row, int col) throws Exception
    {
        row--;
        col--;
        if (row < 0 || row > this.rows) {
            this.throwException (this.E_ROUTB, "row index out of bounds");
        }
        if (col < 0 || col > this.cols) {
            this.throwException (this.E_COUTB, "column index out of bounds");
        }
        Matrix subMatrix = new Matrix (this.rows - 1, this.cols - 1);
        boolean skipRow = false;
        for (int j = 0; j < this.rows - 1; j++) {
            boolean skipCol = false;
            int jSelf = j;
            if (j == row) {
                skipRow = true;
            }
            if (skipRow) {
                jSelf++;
            }
            for (int k = 0; k < this.cols - 1; k++) {
                int kSelf = k;
                if (k == col) {
                    skipCol = true;
                }
                if (skipCol) {
                    kSelf++;
                }
                subMatrix.matrix [j][k] = this.matrix [jSelf][kSelf];
            }
        }
        return subMatrix;
    }

    /**
     * Calculate determinant of matrix
     *
     * @return double The determinant
     *
     * @throw Exception If matrix isn't a square matrix - E_NOSQR
     */
    public double determinant() throws Exception
    {
        if (this.rows != this.cols) {
            this.throwException (this.E_NOSQR, "not a square matrix");
        }
        int n = this.rows;
        double determinant = 0;
        if (n == 1) {
            return this.matrix [0][0];
        } else if (n == 2) {
            return this.matrix [0][0] * this.matrix [1][1] - this.matrix [0][1] * this.matrix [1][0];
        } else {
            for (int i = 0; i < n; i++) {
                double coefficient = Math.pow (-1, i) * this.matrix [0][i];
                Matrix subMatrix = this.excludeRowCol (1, i + 1);
                double subDeterminant = subMatrix.determinant();
                double term = coefficient * subDeterminant;
                determinant += term;
            }
        }
        return determinant;
    }

    /**
     * Calculate the transpose of this matrix
     *
     * @return Matrix The transpose
     */
    public Matrix transpose() throws Exception
    {
        Matrix rslt = new Matrix (this.cols, this.rows);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                rslt.matrix [j][i] = this.matrix [i][j];
            }
        }
        return rslt;
    }

    /**
     * Calculate the cofactor matrix of this matrix
     *
     * @return Matrix The cofactor matrix
     *
     * @throws Exception If matrix isn't a square matrix - E_NOSQR
     */
    public Matrix cofactor() throws Exception
    {
        if (this.rows != this.cols) {
            this.throwException (this.E_NOSQR, "not a square matrix");
        }
        Matrix rslt = new Matrix (this.rows, this.cols);
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                double coefficient = Math.pow (-1, i + 1 + j + 1);
                Matrix subMatrix = this.excludeRowCol (i + 1, j + 1);
                rslt.matrix [i][j] = coefficient * subMatrix.determinant();
            }
        }
        return rslt;
    }

    /**
     * Calculate the adjoint of this matrix
     *
     * @return Matrix The adjoint
     *
     * @throws Exception If matrix isn't a square matrix - E_NOSQR
     */
    public Matrix adjoint() throws Exception
    {
        if (this.rows != this.cols) {
            this.throwException (this.E_NOSQR, "not a square matrix");
        }
        return this.cofactor().transpose();
    }

    /**
     * Calculate the inverse of this matrix
     *
     * @return Matrix The inverse
     *
     * @throws Exception If matrix isn't a square matrix - E_NOSQR
     * @throws Exception If determinant is 0 - E_DETR0
     */
    public Matrix inverse() throws Exception
    {
        if (this.rows != this.cols) {
            this.throwException (this.E_NOSQR, "not a square matrix");
        }
        double det = this.determinant();
        if (det == 0) {
            this.throwException (this.E_DETR0, "determinant is 0");
        }
        Matrix rslt = this.adjoint().scale (1 / this.determinant());
        return rslt;
    }

    /**
     * Display the matrix
     */
    public void print() throws Exception
    {
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                System.out.printf (this.FLOAT_PRECISION + " \t", this.matrix [i][j]);
            }
            System.out.println();
        }
    }

    /**
     * Display the matrix
     *
     * @param message The message to print
     */
    public void print (String message) throws Exception
    {
        System.out.println (message);
        this.print();
    }

    /**
     * Demo function, tests the matrix class
     * Expected output should be:
     * --------------------------
     * Test 1:
     * true
     *
     * Test 2:
     *  0.50   1.50  -2.00 	
     * -1.00   1.00  -1.00 	
     *  0.50  -1.50   2.00
     */
    public static void test() throws Exception
    {
        // Test 1: Given a matrix A
        Matrix A = new Matrix (new double [][] {
            {1, 2, 2},
            {2, 1, 2},
            {2, 2, 1}
        });

        // Prove: A^2 - 4A - 5I(3) = O
        System.out.println ("Test 1:\n" + A.power(2).subtract(A.scale(4)).subtract(Matrix.I(3).scale(5)).equals(Matrix.O(3)));

        // Test 2: Given a matrix B
        Matrix B = new Matrix (new double [][] {
            {1, 0, 1},
            {3, 4, 5},
            {2, 3, 4}
        });

        // find inverse of B
        B.inverse().print("\nTest 2:");
    }
}

File: Matrix.py
class Matrix:

    FLOAT_PRECISION = "%.2f"

    E_0ROWS = "E_0ROWS"    # matrix can't have 0 rows
    E_0COLS = "E_0COLS"    # matrix can't have 0 columns
    E_ROUTB = "E_ROUTB"    # row index out of bounds
    E_COUTB = "E_COUTB"    # column index out of bounds
    E_INCMP = "E_INCMP"    # incompatible matrices
    E_NOSQR = "E_NOSQR"    # not a square matrix
    E_DETR0 = "E_DETR0"    # during inversion, determinant is 0

    def __init__(self, matrix, rows = 0, cols = 0):
        '''
        Create a new Matrix object
        
        Parameters:
            matrix(int[][]): A DDA to initialize, pass None of unknown
            matrix(float[][]): A DDA to initialize, pass None of unknown
            rows(int): If DDA is unknown, pass no. of rows
            cols(int): If DDA is unknown, pass no. of cols

        Raises:
            Error: matrix can't have 0 rows - E_0ROWS
            Error: matrix can't have 0 columns - E_0COLS
        '''
        self.coderr = True
        if rows != 0 and cols != 0:
            self.rows = rows
            self.cols = cols
            self.matrix = [[0 for x in range(cols)] for y in range(rows)]
            return
        self.rows = len(matrix)
        if self.rows == 1:
            matrix = [ matrix ]
        self.cols = len(matrix[0])
        if self.rows < 1:
            self.raise_exception(self.E_0ROWS, "matrix can't have 0 rows")
        if self.cols < 1:
            self.raise_exception(self.E_0COLS, "matrix can't have 0 rows")
        self.matrix = matrix

    def __len__(self):
        return (self.rows, self.cols)

    def __getitem__(self, i):
        return self.matrix[i]

    def __setitem__(self, i, value):
        self.matrix[i] = value

    def raise_exception(self, err_code, err_msg):
        if self.coderr:
            raise Exception(err_code)
        else:
            raise Exception("Matrix: " + err_msg)

    def O(n, cols = 0):
        '''
        Create a null matrix of given size

        Parameters:
            n(int): size of matrix
            cols(int): cols of null matrix(if rows != cols)

        Returns:
            Matrix: a null matrix

        Raises:
            Error: row index out of bounds - E_ROUTB
            Error: column index out of bounds - E_COUTB
        '''
        if cols == 0:
            cols = n
        om = Matrix(None, n, cols)
        return om

    def I(n):
        '''
        Create a unit matrix of given size

        Parameters:
            n(int): size of matrix

        Returns:
            Matrix: a unit matrix

        Raises:
            Error: row index out of bounds - E_ROUTB
            Error: column index out of bounds - E_COUTB
        '''
        im = Matrix(None, n, n)
        for i in range(n):
            im.set(i + 1, i + 1, 1)
        return im

    def get(self, i, j):
        '''
        Get an element of the matrix from an index
        
        Parameters:
            i(int): row wise position of element(starts from 1)
            j(int): column wise position of element(starts from 1)
        
        Returns:
            any: The element

        Raises:
            Error: row index out of bounds - E_ROUTB
            Error: column index out of bounds - E_COUTB
        '''
        i -= 1
        j -= 1
        if i < 0 or i > self.rows:
            self.raise_exception(self.E_ROUTB, "row index out of bounds")
        if j < 0 or j > self.cols:
            self.raise_exception(self.E_COUTB, "column index out of bounds")
        return self[i][j]

    def set(self, i, j, value):
        '''
        Set an element of the matrix to an index
        
        Parameters:
            i(int): row wise position of element(starts from 1)
            j(int): column wise position of element(starts from 1)

        Raises:
            Error: row index out of bounds - E_ROUTB
            Error: column index out of bounds - E_COUTB
        '''
        i -= 1
        j -= 1
        if i < 0 or i > self.rows:
            self.raise_exception(self.E_ROUTB, "row index out of bounds")
        if j < 0 or j > self.cols:
            self.raise_exception(self.E_COUTB, "column index out of bounds")
        self[i][j] = value

    def equals(self, mtb):
        '''
        Compares two matrices for equality

        Parameters:
            mtb(Matrix): The matrix to compare to

        Returns:
            boolean: True if equal
        '''
        if self.rows != mtb.rows or self.cols != mtb.cols:
             return False
        for i in range(self.rows):
            for j in range(self.cols):
                if self[i][j] != mtb[i][j]:
                    return False
        return True

    def __eq__(self, mtb):
        return self.equals(mtb)

    def __contains__(self, el):
        for row in self.m:
            return el in row
        return False

    def add(self, mtb):
        '''
        Adds two compatible matrices

        Parameters:
            mtb(Matrix): The matrix to add

        Returns:
            Matrix: The matrix of sums

        Raises:
            Error: If matrices aren't compatible - E_INCMP
        '''
        if self.rows != mtb.rows or self.cols != mtb.cols:
            self.raise_exception(self.E_INCMP, "incompatible matrices for addition")
        rslt = Matrix(None, self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                rslt[i][j] = self[i][j] + mtb[i][j]
        return rslt

    def __add__(self, mtb):
        return self.add(mtb)

    def subtract(self, mtb):
        '''
        Subtracts two compatible matrices

        Parameters:
            mtb(Matrix): The matrix to subtract

        Returns:
            Matrix: The matrix of differences

        Raises:
            Error: If matrices aren't compatible - E_INCMP
        '''
        if self.rows != mtb.rows or self.cols != mtb.cols:
            self.raise_exception(self.E_INCMP, "incompatible matrices for subtraction")
        rslt = Matrix(None, self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                rslt[i][j] = self[i][j] - mtb[i][j]
        return rslt

    def __sub__(self, mtb):
        return self.subtract(mtb)

    def scale(self, scalar):
        '''
        Multiplies a matrix by a scalar

        Parameters:
            scalar(int): Scalar to multiply by
            scalar(float): Scalar to multiply by

        Returns:
            Matrix: The matrix of products
        '''
        rslt = Matrix(None, self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                rslt[i][j] = scalar * self[i][j]
        return rslt

    def __pos__(self):
        return self

    def __neg__(self):
        return self.scale(-1)

    def multiply(self, mtb):
        '''
        Multiplies two compatible matrices

        Parameters:
            mtb(Matrix): The matrix to multiply by

        Returns:
            Matrix: The matrix after multiplication

        Raises:
            Error: If matrices aren't compatible - E_INCMP
        '''
        if self.cols != mtb.rows:
            self.raise_exception(self.E_INCMP, "incompatible matrices for multiplication")
        m = self.rows
        n = self.cols # same
        n = mtb.rows  # same
        o = mtb.cols
        rslt = Matrix(None, m, o)
        for i in range(m):
            for j in range(o):
                # zeroing the sum without changing datatypes
                sum = self[0][0] - self[0][0]
                for k in range(n):
                    sum += self[i][k] * mtb[k][j]
                rslt[i][j] = sum
        return rslt

    def __mul__(self, nm):
        if isinstance(nm, Matrix):
            return self.multiply(nm)
        elif isinstance(nm, int) or isinstance(nm, float):
            return self.scale(nm)
        else:
            self.raise_exception(self.E_INCMP, "incompatible matrices for multiplication")

    def __rmul__(self, nm):
        return self * nm

    def power(self, index):
        '''
        Calculate matrix to the power of index

        Parameters:
            index(int): Power of matrix

        Returns:
            Matrix: The resulting matrix

        Raises:
            Error: same as errors of Matrix.multiply method
        '''
        if index > 0: tmp = self
        elif index < 0: tmp = self.inverse()
        else: tmp = Matrix.I( min(self.rows, self.cols) )
        result, index = tmp, abs(index)
        for i in range(index - 1):
            result = result.multiply(tmp)
        return result

    def __pow__(self, num):
        return self.power(num)

    def exclude_row_col(self, row, col):
        '''
        Excludes a row and s column and generates a sub matrix. Useful for calculating determinants and cofactor matrices.

        Parameters:
            row(int): The row to exclude(starts from 1)
            col(int): The column to exclude(starts from 1)

        Returns:
            Matrix: The sub matrix

        Raises:
            Error: row index out of bounds - E_ROUTB
            Error: column index out of bounds - E_COUTB
        '''
        row -= 1
        col -= 1
        if row < 0 or row > self.rows:
            self.raise_exception(self.E_ROUTB, "row index out of bounds")
        if col < 0 or col > self.cols:
            self.raise_exception(self.E_COUTB, "column index out of bounds")
        sub_matrix = Matrix(None, self.rows - 1, self.cols - 1)
        skip_row = False
        for j in range(self.rows - 1):
            skip_col = False
            j_self = j
            if j == row:
                skip_row = True
            if skip_row:
                j_self += 1
            for k in range(self.cols - 1):
                k_self = k
                if k == col:
                    skip_col = True
                if skip_col:
                    k_self += 1
                sub_matrix[j][k] = self[j_self][k_self]
        return sub_matrix

    def determinant(self):
        '''
        Calculate determinant of matrix

        Returns:
            int: The determinant
            float: The determinant

        Raises:
            Error: If matrix isn't a square matrix - E_NOSQR
        '''
        if self.rows != self.cols:
            self.raise_exception(self.E_NOSQR, "not a square matrix")
        n = self.rows
        determinant = self[0][0] - self[0][0]
        if n == 1:
            return self[0][0]
        elif n == 2:
            return self[0][0] * self[1][1] - self[0][1] * self[1][0]
        else:
            for i in range(n):
                coefficient = (-1) ** i * self[0][i]
                sub_matrix = self.exclude_row_col(1, i + 1)
                sub_determinant = abs(sub_matrix)
                term = coefficient * sub_determinant
                determinant += term
        return determinant

    def __abs__(self):
        return self.determinant()

    def transpose(self):
        '''
        Calculate the transpose of this matrix

        Returns:
            Matrix: The transpose
        '''
        rslt = Matrix(None, self.cols, self.rows)
        for i in range(self.rows):
            for j in range(self.cols):
                rslt[j][i] = self[i][j]
        return rslt

    def cofactor(self):
        '''
        Calculate the cofactor matrix of this matrix

        Returns:
            Matrix: The cofactor matrix

        Raises:
            Error: If matrix isn't a square matrix - E_NOSQR
        '''
        if self.rows != self.cols:
            self.raise_exception(self.E_NOSQR, "not a square matrix")
        rslt = Matrix(None, rows = self.rows, cols = self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                coefficient = (-1) ** (i + 1 + j + 1)
                sub_matrix = self.exclude_row_col(i + 1, j + 1)
                rslt[i][j] = coefficient * abs(sub_matrix)
        return rslt

    def adjoint(self):
        '''
        Calculate the adjoint of this matrix

        Returns:
            Matrix: The adjoint

        Raises:
            Error: If matrix isn't a square matrix - E_NOSQR
        '''
        if self.rows != self.cols:
            self.raise_exception(self.E_NOSQR, "not a square matrix")
        return self.cofactor().transpose()

    def inverse(self):
        '''
        Calculate the inverse of this matrix

        Returns:
            Matrix: The inverse

        Raises:
            Error: If matrix isn't a square matrix - E_NOSQR
            Error: If determinant is 0 - E_DETR0
        '''
        if self.rows != self.cols:
            self.raise_exception(self.E_NOSQR, "not a square matrix")
        det = abs(self)
        if det == 0:
            self.raise_exception(self.E_DETR0, "determinant is 0")
        rslt =  1 / det * self.adjoint()
        return rslt

    def __invert__(self):
        return self ** -1

    def __str__(self):
        str = ''
        for i in range(self.rows):
            for j in range(self.cols):
                str += ( self.FLOAT_PRECISION % self[i][j] ) + " \t"
            str += "\n"
        return str

    def print(self, message=None):
        '''
        Display the matrix
        '''
        if message != None:
            print(message)
        print(String(self))

    def test():
        '''
        Demo function, tests the matrix class
        Expected output should be:
        --------------------------
        Test 1:
        True

        Test 2:
         0.50   1.50  -2.00 	
        -1.00   1.00  -1.00 	
         0.50  -1.50   2.00
        '''
        # Test 1: Given a matrix A
        A = Matrix([
            [1, 2, 2],
            [2, 1, 2],
            [2, 2, 1]
        ])

        # Prove: A^2 - 4A - 5I(3) = O
        print("Test 1:\n%s" % ( A**2 - 4*A - 5*Matrix.I(3) == Matrix.O(3) ))

        # Test 2: Given a matrix B
        B = Matrix([
            [1, 0, 1],
            [3, 4, 5],
            [2, 3, 4]
        ])

        # find inverse of B
        print("\nTest 2:\n", ~B)

if __name__ == '__main__':
    Matrix.test()
